compiler/hash-ast/src/ast.rs

```diff
@@ -23,15 +29,53 @@ counter! {
     method_visibility: pub,
 }
 
+impl Idx for AstNodeId {
+    fn from_usize(idx: usize) -> Self {
+        Self::from(idx as u32)
+    }
+
+    fn index(self) -> usize {
+        self.0 as usize
+    }
+}
+
+/// The [`SPAN_MAP`] is a global static that is used to store the span
+/// of each AST node. This is used to avoid storing the [Span] on the
+/// [`AstNode<T>`] itself in order for other data structures to be able
+/// to query the [Span] of a node simply by using the [AstNodeId] of the
+/// node.
+static SPAN_MAP: Lazy<RwLock<IndexVec<AstNodeId, Span>>> =
+    Lazy::new(|| RwLock::new(IndexVec::new()));
+
+/// Utilities for working with the [`SPAN_MAP`].
+pub struct SpanMap;
+
+impl SpanMap {
+    /// Get the span of a node by [AstNodeId].
+    pub fn span_of(id: AstNodeId) -> Span {
+        SPAN_MAP.read()[id]
+    }
+
+    /// Get a mutable reference to the [`SPAN_MAP`]. This is only
+    /// internal to the `hash-ast` crate since it creates entries
+    /// in the span map when creating new AST nodes.
+    fn add_span(span: Span) -> AstNodeId {
+        SPAN_MAP.write().push(span)
+    }
+
+    /// Update the span of a node by [AstNodeId].
+    fn update_span(id: AstNodeId, span: Span) {
+        SPAN_MAP.write()[id] = span;
+    }
+}
+
```

Comment:
Why not use the partial_store macro for this?

Reply:
Well its not a partial store, but the single store could be used here. I also wanted to make AstNodeId specifically 4bytes and not 8.

Reply:
Right cause its a total mapping. See comment below on the ID.